config/src/network_id.rs

```diff
@@ -81,12 +81,48 @@ impl NetworkContext {
 /// and handshakes should verify that the NetworkId being used is the same during a handshake,
 /// to effectively ensure communication is restricted to a network.  Network should be checked that
 /// it is not the `DEFAULT_NETWORK`
-#[derive(Clone, Deserialize, Eq, Hash, PartialEq, Serialize)]
+#[derive(Clone, Copy, Eq, Hash, PartialEq, Serialize)]
 #[serde(rename = "NetworkId", rename_all = "snake_case")]
 pub enum NetworkId {
     Validator,
     Public,
-    Private(String),
+    /// An invalid type due to migration from old NetworkId type to new type.  This will *not* deserialize
+    /// TODO: Once fully migrated, this could become Vfn through migration
+    Invalid,
+    Vfn,
+}
+
+impl<'de> Deserialize<'de> for NetworkId {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(rename = "NetworkId", rename_all = "snake_case")]
+        enum ConvertNetworkId {
+            Validator,
+            Public,
+            Private(String),
+            Vfn,
+        }
+
+        // A hack around NetworkId to convert the old type to the new version
+        match ConvertNetworkId::deserialize(deserializer)? {
+            ConvertNetworkId::Validator => Ok(NetworkId::Validator),
+            ConvertNetworkId::Public => Ok(NetworkId::Public),
+            ConvertNetworkId::Vfn => Ok(NetworkId::Vfn),
+            ConvertNetworkId::Private(inner) => {
+                if VFN_NETWORK == inner {
+                    Ok(NetworkId::Vfn)
+                } else {
+                    Err(Error::unknown_variant(
+                        &inner,
+                        &["public", "validator", VFN_NETWORK],
+                    ))
+                }
+            }
+        }
+    }
```

Comment:
We also need to serialize as the old format for old nodes to understand our handshake :'(

Reply:
i guess you could do a 2-step upgrade

(1) today: nodes only read and send Private(String) version
(2) roll out upgrade where nodes use `enum { Validator, Public, _, VFN }` that can deserialize a `enum { Validator, Public, Private(String) }` and a `enum { Validator, Public, _, VFN }` but serializes as `enum { Validator, Public, Private(String) }`
(3) roll out upgrade where nodes only send `enum { Validator, Public, _, VFN }`
(4) eventually clean up

Reply:
or we just keep the same wire format, i.e., `enum { Validator, Public, Private(String) }` but use a compact in-node representation that is also `Copy`

Reply:
I just did the full migration strategy.

Honestly, it doesn't make much of a difference except on serialization of handshakes, and we could change it at our leisure after this change.